### DISCUSSION TITLE
Feedback: Confusion about config syntax for Safari potentially results in reduced support for Safari

### DISCUSSION THREAD

**Discussion by nickserv** [browserslist/browserslist#781] [General]
I've been using various tools that support Browserslist configs for a while now, and was recently discussing adding support for Baseline, when I got this helpful reply from @romainmenke:

> `last 2 safari versions` is `16.5`, `16.4` whereas Baseline has `16` and `15` at this time

At first I thought this was a bug or missing feature. I did some more troubleshooting and research and found the `major versions` syntax for Safari and iOS, which works nicely with `last 2 safari major versions`.

I found this issue difficult to discover because there's no linter warning for it and [browsersl.ist](https://browsersl.ist/) lists so many Safari versions that I didn't think to read them carefully. I suspect other developers may also find this difficult to understand, as popular open source tools like Create React App, Vite, Docusaurus, and Gatsby have made the same mistake, and I found [approximately 9,400 similar configs on GitHub](https://github.com/search?q=content%3A%22safari+versions%22+AND+%28path%3A.browserslistrc+OR+path%3Apackage.json+AND+content%3Abrowserslist%29&type=code&repo=&langOverride=&start_value=1). Without fixing this, a significant portion of Safari users could be left unsupported. In the cause of Docusaurus, I noticed its Browserslist config could potentially cause half of its developer users to be unsupported, according to [browsersl.ist](https://browsersl.ist/).

Personally, I think `versions` should use major versions for Safari and iOS by default, as it appears to do for all other browsers and systems. If this isn't possible, I'm hoping to at least raise awareness and propose DX improvements for Safari support. I also believe Safari version support is more sensitive than other browsers, as it's the only browser engine on iOS and requires system updates, and doesn't automatically update on older macOS versions like evergreen browsers can.

## Related issues
- https://github.com/web-platform-dx/feature-set/issues/264
- https://github.com/facebook/docusaurus/issues/9173
- https://github.com/vitejs/vite/issues/13921

**Comment by romainmenke** [comment]
> Personally, I think versions should use major versions for Safari and iOS, as it appears to do for all other browsers and systems. ` If this isn't possible, I'm hoping to at least raise awareness and propose DX improvements.

This wouldn't be ideal :)

Baseline was using a bad definition where it is thinking in `semver` and was forcing that thinking onto the versioning schemes of all browsers.

A Safari minor version bump is equivalent to a Chrome major version bump.

Safari majors :
- are tied to marketing for new OS versions
- happen once a year
- are ties to larger changes to the browser itself, not the underlying engine

Safari minors :
- happen multiple times a year
- contain new browser engine features

Chrome majors :
- happen multiple times a year
- contain new browser engine features

Firefox majors :
- happen multiple times a year
- contain new browser engine features

Samsung Internet minors :
- happen multiple times a year
- contain new browser engine features

Baseline/WebDX have since corrected their definition (only internally) so that Safari minors are equivalent to Chrome majors. This change isn't yet reflected on marketing pages and blog posts for Baseline

---------------

IMO the baseline definitions are just complete bogus.
They are so wildly incorrect and unhelpful that is best to just ignore Baseline entirely for now.

I can say that here, but it wouldn't be helpful to say that in the WebDX group.
I really hope that their work progresses up to a point where they do have a good definition.

There is a very large chance that that definition will be extremely close to `defaults` in browserslist.

`defaults` in browserslist does what you want and has been tested in the wild.

-------------

It is however critically important that no one starts changing things that work today because of bogus reasoning in Baseline.

It is better to let them work through their mistakes without causing further fallout in the wider dev community :D

**Reply by nickserv** [reply]
> Baseline was using a bad definition where it is thinking in `semver` and was forcing that thinking onto the versioning schemes of all browsers.

I mentioned Baseline primarily to explain how you helped me discover this issue. I'd have the same argument if Baseline didn't exist. Proving this, the same issue seems to be present in Create React App, Vite, Docusaurus, Gatsby, and [thousands of other repositories](https://github.com/search?q=content%3A%22safari+versions%22+AND+%28path%3A.browserslistrc+OR+path%3Apackage.json+AND+content%3Abrowserslist%29&type=code&repo=&langOverride=&start_value=1), many of which precede Baseline.

Additionally, the npm community generally assumes all versions are semver compatible unless described otherwise due to the default behavior of package managers, and projects that don't follow semver (like TypeScript and Prettier) have already updated their documentation to avoid confusion.

> A Safari minor version bump is equivalent to a Chrome major version bump.

This argument is valid only when comparing the contents of updates at a basic level, but it shouldn't be applied to browser version support because Safari handles updates differently from virtually all other browsers. More naunce and statistics are necessary:

### Explanation

Safari being the only modern browser that isn't evergreen (even Android browsers are) means users stay on older versions of Safari without updating more frequently, leading to higher variance in versions. As a result, more Safari versions must be supported to maintain support consistent with other browsers. 

### Technical details

All modern browsers (including Baseline and Android) except Safari install updates without intervention. Safari is part of macOS and iOS, so it needs to be updated through system updates and not the App Store. Additionally, updating Safari on iOS requires rebooting, which delays updates further. Because the iOS App Store bans all browser engines except Safari, this means you cannot get new engine features updates without rebooting on a supported device. Apple also doesn't backport features to old versions of iOS, so you will stop receiving engine features as soon as your device or system is unsupported, which is already an issue for the iPhone 7 (Plus) which no longer receives the latest iOS updates.

### Statistics

I compared the coverage for the last two versions of all modern browsers to all browsers. Note that the standard deviation for Safari majors is lower than for minors. This proves that Safari majors are closer in version support to majors in other browsers.

Browser | last 2 * versions | * > 0 | : | σ
-- |  -- | -- | -- | --
Chrome | 14.6% |  24.6% | 59.3% |  
Edge | 4.3% |  5.2% | 82.7% |  
Firefox | 1.3% |  2.4% | 54.2% |  
Safari | 2.5% |  5.9% | 42.4% | 16.93%
Safari major | 5.0% | 5.9% | 84.7% | 15.73%



> Safari majors are ties to larger changes to the browser itself, not the underlying engine

In my experience, Safari releases most of its platform support in major releases, and minor releases often focus more on security updates (necessary because it's not evergreen). Safari 16 for example has many HTML, CSS, and JS features that have been missing for years, but not major UI changes (which Chrome often has several times per year).

**Reply by ai** [reply]
There is no way to change the current logic of queries because it will be a major changes.

**Reply by nickserv** [reply]
I understand that, but we can still improve DX without major breaking changes. How about adding lint warnings and fixing the `defaults` to include `last 2 major safari versions`? I can make issues or pull requests detailing suggestions if you prefer.

**Reply by ai** [reply]
We don't have `add FF esr` warning. Why we should ask to add 2 major Safari in this case.

Browserslist idea is to promote browsers diversity and people access to the internet.

Apple users as I understand can always update to latest Safari. So as I understand this change will not help is in our tasks.

**Reply by nickserv** [reply]
> We don't have `add FF esr` warning. Why we should ask to add 2 major Safari in this case.

Firefox ESR has extremely low usage compared to Firefox and Safari. It also has different update semantics than Safari, so it makes sense to treat them differently.

> Browserslist idea is to promote browsers diversity and people access to the internet.

I believe supporting Safari users consistently should be part of this goal.

> Apple users as I understand can always update to latest Safari. So as I understand this change will not help is in our tasks.

Reasons users don't update Safari:

1. Safari updates are only installable on supported Apple devices. Popular devices like the iPhone 7 (Plus) are already incompatible with these Browserlist queries because they don't receive Safari updates anymore.
2. iOS Safari updates are bundled with other iOS updates. Devices need to have plenty of disk space, be rebooted, and then be unavailable for several minutes until the update has been installed. Many users delay or disable updates to avoid this.
3. Some enterprise policies may prevent installing updates.
4. Safari doesn't automatically check for updates at launch and remind users to restart it to install updates, which most other browsers do.

**Reply by ai** [reply]
Does iPhone 7 has security updates?

**Reply by nickserv** [reply]
Yes, but they're not guaranteed, and they don't seem to backport any new engine features to Safari.

**Reply by ai** [reply]
We have a `>0.5%` query for popular cases.

As I see old Safari is already in `defaults`.

**Reply by nickserv** [reply]
This still excludes about 35% of Safari 15 users. I think `last 2 safari major versions` is closer to the expectations users have about support for other browsers, and it's also closer to the market shares of other browsers statistically (see [above](https://github.com/orgs/browserslist/discussions/781#discussioncomment-6516951)) .

**Reply by romainmenke** [reply]
> This still excludes about 35% of Safari 15 users. I think last 2 safari major versions is closer to the expectations users have about support for other browsers, and it's also closer to the market shares of other browsers statistically

I think you are confusing two different things.

`last N Safari` queries aren't intended to reflect the majority of users.
They must return the correct **meaningful** versions for Safari.

Because of how Safari is versioned these **must** include minor versions.

Changing this would be massively breaking and would upset WebKit/Apple and Samsung Internet.

If you want a query that reflects a majority of users you **must** use `>n%` queries.

**Reply by romainmenke** [reply]
Please see : https://github.com/web-platform-dx/feature-set/issues/173#issuecomment-1555386336

Apple/WebKit are very vocal about minor versions being equivalent to major versions in Chrome/Firefox.

The same is true for Samsung Internet.

**Reply by nickserv** [reply]
> Changing this would be massively breaking and would upset WebKit/Apple and Samsung Internet.
>
> If you want a query that reflects a majority of users you **must** use `>n%` queries.

There's nothing that prevents using `>n%` queries if more coverage is needed, but this isn't an argument against incremental improvements to potentially confusing queries. And there doesn't have to be abny breaking change.

**Reply by romainmenke** [reply]
> And there doesn't have to be abny breaking change.

It is a breaking change :)

For example I depend on `browserslist` in `cssdb`, `postcss-preset-env`, `core-web` and `polyfill.io` to construct and verify support data.

Changing the versions that are returned for such queries would break either those tools directly or glue scripts that support them.

---------

But more importantly it is still an incorrect assumption that Safari majors are their meaningful releases. They are not.

This is what I have been spending all my free time on for the last 4 years.  
Please consider the possibility that the current situation might be correct :)

**Reply by nickserv** [reply]
> It is a breaking change :)

No it's not: https://github.com/orgs/browserslist/discussions/781#discussioncomment-6517004

**Reply by romainmenke** [reply]
I am leaving this conversation.

I am a core maintainer of large projects for which this is going to be a breaking change.
I have given you extra information from good sources.

I don't want to debate what is or isn't true when I am simply presenting you with facts.
